File: prolanlab-java/src/main/java/com/prolanlab/tutorials/python/PythonTutorialController.java
package com.prolanlab.tutorials.python;

import com.prolanlab.tutorials.python.fundamentals.PythonFundamentals;
import com.prolanlab.tutorials.python.controlstructures.PythonControlStructures;
import com.prolanlab.tutorials.python.flowcontrol.PythonFlowControl;
import com.prolanlab.tutorials.python.classesandoop.PythonClassesAndOOP;
import com.prolanlab.tutorials.python.datastructures.PythonDataStructures;
import com.prolanlab.Main;

import java.util.Scanner;

/**
 * Python Tutorial Controller - Main controller for Python tutorials
 */
public class PythonTutorialController {
    private final Scanner scanner = Main.getScanner();
    private final PythonFundamentals fundamentals = new PythonFundamentals();
    private final PythonControlStructures controlStructures = new PythonControlStructures();
    private final PythonFlowControl flowControl = new PythonFlowControl();
    private final PythonClassesAndOOP classesAndOOP = new PythonClassesAndOOP();
    private final PythonDataStructures dataStructures = new PythonDataStructures();
    
    public void start() {
        boolean inPythonTutorial = true;
        
        while (inPythonTutorial) {
            displayPythonMenu();
            int choice = getUserChoice();
            
            switch (choice) {
                // Fundamentals (1-15)
                case 1: fundamentals.showPythonIntro(); break;
                case 2: fundamentals.showPythonSyntax(); break;
                case 3: fundamentals.showPythonStatements(); break;
                case 4: fundamentals.showPythonPrintText(); break;
                case 5: fundamentals.showPythonPrintNumbers(); break;
                case 6: fundamentals.showPythonComments(); break;
                case 7: fundamentals.showPythonVariables(); break;
                case 8: fundamentals.showPythonMultipleValues(); break;
                case 9: fundamentals.showPythonOutputVariables(); break;
                case 10: fundamentals.showPythonGlobalVariables(); break;
                case 11: fundamentals.showPythonIdentifiers(); break;
                case 12: fundamentals.showPythonConstants(); break;
                case 13: fundamentals.showPythonDataTypes(); break;
                case 14: fundamentals.showPythonNumbers(); break;
                case 15: fundamentals.showPythonTypeCasting(); break;
                
                // Control Structures (16-32)
                case 16: controlStructures.showPythonOperators(); break;
                case 17: controlStructures.showPythonArithmetic(); break;
                case 18: controlStructures.showPythonAssignment(); break;
                case 19: controlStructures.showPythonComparison(); break;
                case 20: controlStructures.showPythonLogical(); break;
                case 21: controlStructures.showPythonIdentity(); break;
                case 22: controlStructures.showPythonMembership(); break;
                case 23: controlStructures.showPythonBitwise(); break;
                case 24: controlStructures.showPythonPrecedence(); break;
                case 25: controlStructures.showPythonStrings(); break;
                case 26: controlStructures.showPythonSlicing(); break;
                case 27: controlStructures.showPythonConcatenation(); break;
                case 28: controlStructures.showPythonFormatStrings(); break;
                case 29: controlStructures.showPythonEscapeChars(); break;
                case 30: controlStructures.showPythonStringMethods(); break;
                case 31: controlStructures.showPythonMath(); break;
                case 32: controlStructures.showPythonBooleans(); break;
                
                // Flow Control & Collections (33-42)
                case 33: flowControl.showPythonIfElse(); break;
                case 34: flowControl.showPythonWhileLoop(); break;
                case 35: flowControl.showPythonForLoop(); break;
                case 36: flowControl.showPythonNestedLoops(); break;
                case 37: flowControl.showPythonForEach(); break;
                case 38: flowControl.showPythonBreakContinue(); break;
                case 39: flowControl.showPythonTuples(); break;
                case 40: flowControl.showPythonLists(); break;
                case 41: flowControl.showPythonSets(); break;
                case 42: flowControl.showPythonDictionaries(); break;
                
                // Classes & OOP (43-51)
                case 43: classesAndOOP.showPythonOOP(); break;
                case 44: classesAndOOP.showPythonClassesObjects(); break;
                case 45: classesAndOOP.showPythonSelfParameter(); break;
                case 46: classesAndOOP.showPythonMethods(); break;
                case 47: classesAndOOP.showPythonProperties(); break;
                case 48: classesAndOOP.showPythonInheritance(); break;
                case 49: classesAndOOP.showPythonPolymorphism(); break;
                case 50: classesAndOOP.showPythonEncapsulation(); break;
                case 51: classesAndOOP.showPythonInnerClasses(); break;
                
                // Data Structures (52-60)
                case 52: dataStructures.showPythonListsArrays(); break;
                case 53: dataStructures.showPythonStacks(); break;
                case 54: dataStructures.showPythonQueues(); break;
                case 55: dataStructures.showPythonLinkedLists(); break;
                case 56: dataStructures.showPythonHashTables(); break;
                case 57: dataStructures.showPythonTrees(); break;
                case 58: dataStructures.showPythonBinaryTrees(); break;
                case 59: dataStructures.showPythonGraphs(); break;
                case 60: dataStructures.showPythonAlgorithms(); break;
                
                case 61:
                    System.out.println("Returning to tutorial menu...");
                    inPythonTutorial = false;
                    break;
                default:
                    System.out.println("Invalid option. Please try again.");
            }
            
            if (inPythonTutorial && choice != 61) {
                System.out.println("\n" + "=".repeat(75));
                System.out.println("Press Enter to return to Python tutorial menu...");
                scanner.nextLine();
            }
        }
    }
    
    private void displayPythonMenu() {
        System.out.println("\n" + "=".repeat(75));
        System.out.println("                    [P] PYTHON TUTORIAL");
        System.out.println("=".repeat(75));
        System.out.println(">> FUNDAMENTALS");
        System.out.println(" 1. Intro                  2. Syntax");
        System.out.println(" 3. Statements             4. Print Text");
        System.out.println(" 5. Print Numbers          6. Comments");
        System.out.println(" 7. Variables              8. Multiple Values");
        System.out.println(" 9. Output Variables      10. Global Variables");
        System.out.println("11. Identifiers           12. Constants");
        System.out.println("13. Data Types            14. Numbers");
        System.out.println("15. Type Casting");
        System.out.println();
        System.out.println(">> CONTROL STRUCTURES");
        System.out.println("16. Operators             17. Arithmetic");
        System.out.println("18. Assignment            19. Comparison");
        System.out.println("20. Logical               21. Identity");
        System.out.println("22. Membership            23. Bitwise");
        System.out.println("24. Precedence            25. Strings");
        System.out.println("26. Slicing               27. Concatenation");
        System.out.println("28. Format Strings        29. Escape Chars");
        System.out.println("30. String Methods        31. Math");
        System.out.println("32. Booleans");
        System.out.println();
        System.out.println(">> FLOW CONTROL & COLLECTIONS");
        System.out.println("33. If..Else              34. While Loop");
        System.out.println("35. For Loop              36. Nested Loops");
        System.out.println("37. For-Each              38. Break/Continue");
        System.out.println("39. Tuples                40. Lists");
        System.out.println("41. Sets                  42. Dictionaries");
        System.out.println();
        System.out.println(">> CLASSES & OOP");
        System.out.println("43. OOP                   44. Classes/Objects");
        System.out.println("45. Self Parameter        46. Methods");
        System.out.println("47. Properties            48. Inheritance");
        System.out.println("49. Polymorphism          50. Encapsulation");
        System.out.println("51. Inner Classes");
        System.out.println();
        System.out.println(">> DATA STRUCTURES");
        System.out.println("52. Lists and Arrays      53. Stacks");
        System.out.println("54. Queues                55. Linked Lists");
        System.out.println("56. Hash Tables           57. Trees");
        System.out.println("58. Binary Trees          59. Graphs");
        System.out.println("60. Algorithms");
        System.out.println();
        System.out.println("61. << Back to Tutorial Menu");
        System.out.println("=".repeat(75));
        System.out.print("Enter your choice (1-61): ");
    }
    
    private int getUserChoice() {
        try {
            int choice = scanner.nextInt();
            scanner.nextLine(); // Consume the newline
            return choice;
        } catch (Exception e) {
            scanner.nextLine(); // Clear invalid input
            return -1;
        }
    }
}

File: prolanlab-python/src/prolanlab/tutorials/python/python_tutorial_controller.py
"""
Python Tutorial Controller - Main controller for Python tutorials
"""

from prolanlab.tutorials.python.fundamentals.python_fundamentals import PythonFundamentals
from prolanlab.tutorials.python.controlstructures.python_control_structures import PythonControlStructures
from prolanlab.tutorials.python.flowcontrol.python_flow_control import PythonFlowControl
from prolanlab.tutorials.python.classesandoop.python_classes_and_oop import PythonClassesAndOOP
from prolanlab.tutorials.python.datastructures.python_data_structures import PythonDataStructures


class PythonTutorialController:
    def __init__(self):
        self.fundamentals = PythonFundamentals()
        self.control_structures = PythonControlStructures()
        self.flow_control = PythonFlowControl()
        self.classes_and_oop = PythonClassesAndOOP()
        self.data_structures = PythonDataStructures()

    def start(self):
        in_python_tutorial = True
        
        while in_python_tutorial:
            self._display_python_menu()
            choice = self._get_user_choice()
            
            # Fundamentals (1-15)
            if choice == 1:
                self.fundamentals.show_python_intro()
            elif choice == 2:
                self.fundamentals.show_python_syntax()
            elif choice == 3:
                self.fundamentals.show_python_statements()
            elif choice == 4:
                self.fundamentals.show_python_print()
            elif choice == 5:
                self.fundamentals.show_python_numbers()
            elif choice == 6:
                self.fundamentals.show_python_comments()
            elif choice == 7:
                self.fundamentals.show_python_variables()
            elif choice == 8:
                self.fundamentals.show_python_multiple_values()
            elif choice == 9:
                self.fundamentals.show_python_output_variables()
            elif choice == 10:
                self.fundamentals.show_python_global_local()
            elif choice == 11:
                self.fundamentals.show_python_identifiers()
            elif choice == 12:
                self.fundamentals.show_python_constants()
            elif choice == 13:
                self.fundamentals.show_python_data_types()
            elif choice == 14:
                self.fundamentals.show_python_number_types()
            elif choice == 15:
                self.fundamentals.show_python_type_casting()
                
            # Control Structures (16-32)
            elif choice == 16:
                self.control_structures.show_python_operators()
            elif choice == 17:
                self.control_structures.show_python_arithmetic()
            elif choice == 18:
                self.control_structures.show_python_assignment()
            elif choice == 19:
                self.control_structures.show_python_comparison()
            elif choice == 20:
                self.control_structures.show_python_logical()
            elif choice == 21:
                self.control_structures.show_python_identity()
            elif choice == 22:
                self.control_structures.show_python_membership()
            elif choice == 23:
                self.control_structures.show_python_bitwise()
            elif choice == 24:
                self.control_structures.show_python_precedence()
            elif choice == 25:
                self.control_structures.show_python_strings()
            elif choice == 26:
                self.control_structures.show_python_slicing()
            elif choice == 27:
                self.control_structures.show_python_concatenation()
            elif choice == 28:
                self.control_structures.show_python_format_strings()
            elif choice == 29:
                self.control_structures.show_python_escape_chars()
            elif choice == 30:
                self.control_structures.show_python_string_methods()
            elif choice == 31:
                self.control_structures.show_python_math_module()
            elif choice == 32:
                self.control_structures.show_python_booleans()
                
            # Flow Control (33-42)
            elif choice == 33:
                self.flow_control.show_python_conditions()
            elif choice == 34:
                self.flow_control.show_python_while()
            elif choice == 35:
                self.flow_control.show_python_for()
            elif choice == 36:
                self.flow_control.show_python_nested_loops()
            elif choice == 37:
                self.flow_control.show_python_for_each()
            elif choice == 38:
                self.flow_control.show_python_break_continue()
            elif choice == 39:
                self.flow_control.show_python_tuples()
            elif choice == 40:
                self.flow_control.show_python_lists()
            elif choice == 41:
                self.flow_control.show_python_sets()
            elif choice == 42:
                self.flow_control.show_python_dictionaries()
                
            # Classes & OOP (43-51)
            elif choice == 43:
                self.classes_and_oop.show_oop_principles()
            elif choice == 44:
                self.classes_and_oop.show_python_classes()
            elif choice == 45:
                self.classes_and_oop.show_python_self()
            elif choice == 46:
                self.classes_and_oop.show_python_methods()
            elif choice == 47:
                self.classes_and_oop.show_python_properties()
            elif choice == 48:
                self.classes_and_oop.show_python_inheritance()
            elif choice == 49:
                self.classes_and_oop.show_python_polymorphism()
            elif choice == 50:
                self.classes_and_oop.show_python_encapsulation()
            elif choice == 51:
                self.classes_and_oop.show_python_nested_classes()
                
            # Advanced & Collections (52-61)
            elif choice == 52:
                self.data_structures.show_python_list_advanced()
            elif choice == 53:
                self.data_structures.show_python_dict_advanced()
            elif choice == 54:
                self.data_structures.show_python_set_advanced()
            elif choice == 55:
                self.data_structures.show_python_tuple_advanced()
            elif choice == 56:
                self.data_structures.show_python_comprehensions()
            elif choice == 57:
                self.data_structures.show_python_generators()
            elif choice == 58:
                self.data_structures.show_python_iterators()
            elif choice == 59:
                self.data_structures.show_python_decorators()
            elif choice == 60:
                self.data_structures.show_python_error_handling()
            elif choice == 61:
                self.data_structures.show_python_modules()
                
            elif choice == 62:
                print("Returning to tutorial menu...")
                in_python_tutorial = False
            else:
                print("Invalid option. Please try again.")
                
            if in_python_tutorial and choice != 62:
                print("\nPress Enter to continue...")
                input()

    def _display_python_menu(self):
        print("\n" + "=" * 75)
        print("                   [PYTHON] TUTORIALS")
        print("=" * 75)
        print("\n【FUNDAMENTALS】")
        print(" 1. Python Introduction & Philosophy   9. Variable Output Techniques")
        print(" 2. Syntax & Indentation Rules        10. Global vs Local Scope")
        print(" 3. Statements & Execution            11. Identifiers & Naming")
        print(" 4. Print Functions                   12. Constants & Immutability")
        print(" 5. Number Output & Formatting        13. Data Types Overview")
        print(" 6. Comments & Documentation          14. Number Types (int, float)")
        print(" 7. Variables & Assignment            15. Type Casting & Conversion")
        print(" 8. Multiple Value Assignments")
        print("\n【CONTROL STRUCTURES】")
        print("16. Operators Overview               25. String Creation")
        print("17. Arithmetic Operations            26. String Slicing")
        print("18. Assignment Operators             27. String Concatenation")
        print("19. Comparison Operators             28. Format Strings (f-strings)")
        print("20. Logical Operators                29. Escape Characters")
        print("21. Identity Operators               30. String Methods")
        print("22. Membership Operators             31. Math Module & Functions")
        print("23. Bitwise Operations               32. Boolean Operations")
        print("24. Operator Precedence")
        print("\n【FLOW CONTROL & COLLECTIONS】")
        print("33. Conditional Statements           39. Tuples & Immutable Sequences")
        print("34. While Loops & Control            40. Lists & Mutable Sequences")
        print("35. For Loops & Iteration            41. Sets & Unique Collections")
        print("36. Nested Loop Structures           42. Dictionaries & Key-Value")
        print("37. For-each Loop Patterns")
        print("38. Break & Continue Statements")
        print("\n【CLASSES & OOP】")
        print("43. OOP Principles                   48. Inheritance & super()")
        print("44. Classes & Objects               49. Polymorphism & Duck Typing")
        print("45. Self Parameter                  50. Encapsulation & Private")
        print("46. Methods (instance, class)       51. Inner/Nested Classes")
        print("47. Properties & Decorators")
        print("\n【ADVANCED TOPICS】")
        print("52. Advanced Lists                  58. Iterators & Protocols")
        print("53. Advanced Dictionaries           59. Decorators & Functions")
        print("54. Advanced Sets                   60. Error Handling & Exceptions")
        print("55. Advanced Tuples                 61. Modules & Packages")
        print("56. List/Dict Comprehensions")
        print("57. Generators & yield")
        print()
        print("62. << Back to Tutorial Menu")
        print("=" * 75)
        print("Enter your choice (1-62): ", end="")

    def _get_user_choice(self):
        try:
            choice = int(input())
            return choice
        except ValueError:
            return -1